File: 2015/17/17.py
from itertools import combinations

def parser(data: str):
    return tuple(int(container) for container in data.splitlines())

def solver(containers: list[int]):
    fills: list[int] = list(sum(sum(comb) == 150 for comb in combinations(containers, size)) for size in range(len(containers)))
    fills = list(fill for fill in fills if fill)
    yield fills[0]
    yield sum(fills)

File: 2015/17/17.rs
aoc::main!();

fn parser(input: &str) -> Vec<usize> {
    input
    .lines()
    .map(|n| n.parse::<usize>().unwrap())
    .collect_vec()
}

fn part_1(containers : Vec<usize>) -> usize {
    solver(containers)
    .iter()
    .sum()

}

fn part_2(containers : Vec<usize>) -> usize {
    *solver(containers)
    .iter()
    .find_or_first(|s| **s != 0)
    .unwrap()
}

fn solver(containers : Vec<usize>) -> Vec<usize> {
    (3..=containers.len())
    .map(|i|
        containers
        .iter()
        .combinations(i)
        .filter(|c| c.iter().fold(0, |acc,b| acc + **b ) == 150)
        .count())
        .collect_vec()
}